src/Workspaces/Core/Portable/Editing/ImportAdderService.cs

```diff
@@ -244,11 +214,32 @@ private SyntaxNode GenerateNamespaceImportDeclaration(INamespaceSymbol namespace
                 }
             }
 
-            // we don't add simplifier annotations here, 
-            // since whatever added the symbol annotation probably also added simplifier annotations,
-            // and if not they probably didn't for a reason
+            if (importsToAdd.Count == 0)
+                return document;
+
+            var context = first == null || last == null ? null : first.GetCommonRoot(last);
 
-            return (importsToAdd.ToImmutableAndFree(), addedSymbols, nodes is var (first, last) ? first.GetCommonRoot(last) : null, root);
+            {
+                // Mark the context with an annotation. 
+                // This will allow us to find it after we have called MakeSafeToAddNamespaces.
+                var annotation = new SyntaxAnnotation();
+                RoslynDebug.Assert(context is object);
+                document = document.WithSyntaxRoot(root.ReplaceNode(context, context.WithAdditionalAnnotations(annotation)));
+                root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                model = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+                // Make Safe to add namespaces
+                document = document.WithSyntaxRoot(
+                   MakeSafeToAddNamespaces(root, addedSymbols, model, document.Project.Solution.Workspace, cancellationToken));
+                root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                model = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+                // Find the context. It might be null if we have removed the context in the process of complexifying the tree.
+                context = root.DescendantNodesAndSelf().FirstOrDefault(x => x.HasAnnotation(annotation)) ?? root;
+            }
+
+            root = addImportsService.AddImports(model.Compilation, root, context, importsToAdd, generator, placeSystemNamespaceFirst, cancellationToken);
+            return document.WithSyntaxRoot(root);
```

Comment:
this is just a move of the code that used to be in the caller.  it will go away in a following commit.

Reply:
It did not seem to go away? :smile:

Reply:
weir placement.  ignore this comment.  what happened is i both merged ode in one commit then extracted later.